ocaml/typing/env.ml

```diff
@@ -2349,79 +2349,75 @@ let add_region_lock env =
-      map, add_type ~check:false ?shape id decl env
-  | Sig_typext(id, ext, _, _) ->
-      let map, shape = proj_shape (map, mod_shape) (Shape.Item.extension_constructor id) in
-      map, add_extension ~check:false ?shape ~rebind:false id ext env
-  | Sig_module(id, presence, md, _, _) ->
-      let map, shape = proj_shape (map, mod_shape) (Shape.Item.module_ id) in
-      map, add_module_declaration ~check:false ?shape id presence md env
-  | Sig_modtype(id, decl, _)  ->
-      let map, shape = proj_shape (map, mod_shape) (Shape.Item.module_type id) in
-      map, add_modtype ?shape id decl env
-  | Sig_class(id, decl, _, _) ->
-      let map, shape = proj_shape (map, mod_shape) (Shape.Item.class_ id) in
-      map, add_class ?shape id decl env
-  | Sig_class_type(id, decl, _, _) ->
-      let map, shape = proj_shape (map, mod_shape) (Shape.Item.class_type id) in
-      map, add_cltype ?shape id decl env
-
-let rec add_signature (map, mod_shape) sg env =
-  match sg with
-      [] -> map, env
-  | comp :: rem ->
-      let map, env = add_item (map, mod_shape) comp env in
-      add_signature (map, mod_shape) rem env
-
-let add_item_lazy (map, mod_shape) comp env =
-  let open Subst.Lazy in
-  match comp with
-  | SigL_value(id, decl, _) ->
-      let map, shape = proj_shape (map, mod_shape) (Shape.Item.value id) in
-      map, add_value ?shape id decl env
-  | SigL_type(id, decl, _, _) ->
-      let map, shape = proj_shape (map, mod_shape) (Shape.Item.type_ id) in
-      map, add_type ~check:false ?shape id decl env
-  | SigL_typext(id, ext, _, _) ->
-      let map, shape = proj_shape (map, mod_shape) (Shape.Item.extension_constructor id) in
-      map, add_extension ~check:false ?shape ~rebind:false id ext env
-  | SigL_module(id, presence, md, _, _) ->
-      let map, shape = proj_shape (map, mod_shape) (Shape.Item.module_ id) in
-      map, add_module_declaration_lazy ~update_summary:true ?shape id presence md env
-  | SigL_modtype(id, decl, _)  ->
-      let map, shape = proj_shape (map, mod_shape) (Shape.Item.module_type id) in
-      map, add_modtype_lazy ~update_summary:true ?shape id decl env
-  | SigL_class(id, decl, _, _) ->
-      let map, shape = proj_shape (map, mod_shape) (Shape.Item.class_ id) in
-      map, add_class ?shape id decl env
-  | SigL_class_type(id, decl, _, _) ->
-      let map, shape = proj_shape (map, mod_shape) (Shape.Item.class_type id) in
-      map, add_cltype ?shape id decl env
-
-let rec add_signature_lazy (map, mod_shape) sg env =
-  match sg with
-      [] -> map, env
-  | comp :: rem ->
-      let map, env = add_item_lazy (map, mod_shape) comp env in
-      add_signature_lazy (map, mod_shape) rem env
+module Add_signature(M : sig
+  include Types.S
+  val add_module_declaration: ?arg:bool -> ?shape:Shape.t -> check:bool
+    -> Ident.t -> module_presence -> module_declaration -> t -> t
+  val add_modtype: ?shape:Shape.t -> Ident.t -> modtype_declaration -> t -> t
+end) = struct
```

Comment:
This might look nicer taking a separate `Types.S` argument rather than combining it with the `add` functions into a single argument.